modules/translation/translation.go

```diff
@@ -74,12 +98,17 @@ func InitLocales() {
 	}
 
 	matcher = language.NewMatcher(supportedTags)
+
+	// Make sure en-US is always the first in the slice.
+	setting.Names = moveToFront("English", setting.Names)
+	setting.Langs = moveToFront("en-US", setting.Langs)
```

Comment:
I think we shouldn't do this.

The document has declared that the first language will be used as the default language.

https://docs.gitea.io/en-us/config-cheat-sheet/#i18n-i18n

Reply:
I still prefer a stable locale initialization logic:

1.  add the base language
2. add other languages

Then the `settings.Lans` can be left as it is.

Reply:
Or we should revert these afterwards or duplicate the code to just get the keys/sections of en-US language.

Reply:
I have some ideas about the refactoring, will propose a PR soon.

ps: I think we could decouple the `TrOffset` from the global `DefaultLocales`, instead, make each locale has its own fallback, will try on it.

Reply:
> ps: I think we could also decouple the `TrOffset` from the global `DefaultLocales`, instead, make each locale has its own fallback, will try on it.

Just make sure we don't end up again with 25Mb of memory 😄 otherwise this PR is just a silly refactor with no benefit.

Reply:
Sure, let me try 😊 I also like saving memory

Reply:
https://github.com/Gusted/gitea/pull/4

Some thoughts:
1. "For development, live-reload of the translation files is important.", no need to worry about it, because there was no live-reload before 😂  after the refactoring, the live-reload could be added as a new feature.
2. use a key-string -> int -> translation-text  mapping, to save the memory for duplicate keys.
3. let the locale store to handle the default language fallback

I haven't tested the memory usage, just have a feeling it might be the same.

Reply:
> 1. About "For development, live-reload of the translation files is important.", no need to worry about it, because there was no live-reload before joy  after the refactoring, the live-reload could be added as a new feature.

Yup, definitely for another PR, before this gets too complicated.

> 2\. use a key-string -> int -> translation-text  mapping, to save the memory for duplicate keys. I think that's the minimal memory usage we can achieve now: one copy of keys (as the `textIdxMap` key), one copy of all translations, they are connected by `int` keys (unless there would be some more aggressive text compressing algorithm, which seems too complex for such case)

We now end up making a map for each locale. It could be improved, but that's something to figure when I do the memory testing if we need to optimize that away to either one map(somehow) or use PHF(with more extensive comments).